### DISCUSSION TITLE
of::filesystem, ideas

### DISCUSSION THREAD

**Discussion by dimitre** [openframeworks/openFrameworks#7042] [Ideas]
Some ideas
- [ ] of::filesystem::path can be used in more places substituting strings when dealing with paths and filenames
both in OF and PG.
- [ ] Boost can be updated so we can use filesystem v4 which is more similar to std::filesystem
- [ ] we can think about deprecating some old functions as addTrailingSlash and others

--- Past Stuff
Complete PR here https://github.com/openframeworks/openFrameworks/pull/7110

I was wondering if instead of trying to insert filesystem namespace inside of std, how about we just use the 

in ofConstants.h, one of this depending on defines
```c++
namespace filesystem = std::__fs::filesystem;
namespace filesystem = std::fs::filesystem;
namespace filesystem = boost::filesystem;
namespace filesystem = std::experimental::filesystem;
```

in ofFileUtils.h for example
```
bool open(const filesystem::path & path, Mode mode=ReadOnly, bool binary=true);
```

of course I don't understand about std::filesystem so it can be a very bad idea, but would like to know your opinion

**Comment by dimitre** [comment]
https://github.com/DeadSix27/waifu2x-converter-cpp/issues/251#issuecomment-674443196

**Comment by ofTheo** [comment]
Hey @dimitre 
That could definitely be a simpler / cleaner approach and similar to how other projects handle it.

Would be up for trying this as the current approach has some edge cases which cause ambiguous errors.

**Reply by dimitre** [reply]
Nice, I can make a PR soon for testing. Any suggestions of namespace? we can try as the example above?

**Comment by dimitre** [comment]
Other suggestion : probabaly with addons breaking potential, but I think it worths:
there are a lot of functions that manipulate paths, ex: ofToDataPath which converts back and forth from fs::path to string
My suggestion is we migrate to fs::path when possible.

I had to modify my own addons to use ofToDataPath().string()

**Comment by dimitre** [comment]
First draft
https://github.com/dimitre/openFrameworks/tree/lite

**Comment by ofTheo** [comment]
The only downside I can see to this approach:

```
bool open(const filesystem::path & path, Mode mode=ReadOnly, bool binary=true);
``` 

is if someone puts `using namespace std;` then there might be some ambiguous errors if they have forced the `boost::filesystem` to be `filesystem`.  as filesystem could refer either to the boost one or `std::filesystem`

Alternative: 

`namespace fs = std::filesystem;`

with this usage

`bool open(const fs::path & path, Mode mode=ReadOnly, bool binary=true);`

would be safer but it also feels much less clear. 

`bool open(const ofFileSystem::path & path, Mode mode=ReadOnly, bool binary=true);`

could be an option but that sort of hides what it is. I guess the upside would be that it is discoverable in the OF code completion - but again might be confusing? 

One more that might be cleaner and lean into the of namespace we have been talking about:

`bool open(const of::filesystem::path & path, Mode mode=ReadOnly, bool binary=true);`

Curious what people think.

**Comment by 2bbb** [comment]
I agree with `bool open(const of::filesystem::path & path, Mode mode=ReadOnly, bool binary=true);`.

**Comment by dimitre** [comment]
Yeah of::filesystem is nice. I can make some changes in the fork.

**Comment by dimitre** [comment]
About this same subject, is this the right time to remove boost::filesystem ?

**Comment by dimitre** [comment]
PR is now complete and on master, this is great, and it opens new possibilities of rethinking (simplifying, improving) pieces of OF like filePath object and projectGenerator. 
There are some discussion and decisions made on the PR so I'm linking here as a reference https://github.com/openframeworks/openFrameworks/pull/7110

and I'll link this other one as some future ideas arised there too - https://github.com/openframeworks/openFrameworks/pull/7368

**Reply by dimitre** [reply]
Other PR to have in mind - https://github.com/openframeworks/openFrameworks/pull/7361

**Comment by dimitre** [comment]
Lots of things were decided and merged after this discussion, happy to see filesystem in more places. projectGenerator now uses it extensively.
I'm thinking about this next steps, in this specific order:
- [ ] https://github.com/openframeworks/openFrameworks/pull/7383#issuecomment-1823611571
- [ ] https://github.com/openframeworks/openFrameworks/pull/7494
- [ ] https://github.com/openframeworks/openFrameworks/pull/7435

I'll be buying a second hand windows box just to test the windows related PRs and projectGenerator
